accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,188 @@
+
+For scenarios which continue to use username + password, we should encourage users to pass around a `NetworkCredential` object instead. The NetworkCredential class abstracts away how the password is represented in memory: it can be a string, a legacy SecureString, or any future mechanism we create. Encouraging users to pass around this container rather than a raw SecureString instance allows their code to take advantage of future improvements we make to this space, even while they continue to use username + password credentials.
+
+### Preventing secrets from getting into crash dumps
+
+Users might rely on the fact that the SecureString backing memory is enciphered (only on Windows) to prevent secrets from getting into crash dumps. This is not ideal for a variety of reasons. The secret won't be protected before it's converted into a SecureString or after it's converted from a SecureString (such as during a p/invoke call). `SocketsHttpHandler` [converts SecureString instances to normal strings](https://github.com/dotnet/runtime/blob/0e23ef47ff8a2de6a5b5898a849ba00303aba8df/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs#L176-L185) as part of its operation. WPF's `PasswordBox` control [contains contorted code](https://github.com/dotnet/wpf/blob/53812c590c5f140297ed96ee2a78d05dffa7be35/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PasswordTextContainer.cs#L74-L97) to manipulate SecureString instances from string-based input. `SqlClient` [copies SecureString data into an unpinned buffer](https://github.com/dotnet/SqlClient/blob/ead806e87a15be9f54c99db2ebe8918025c20fc9/src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/Interop/SNINativeMethodWrapper.cs#L923) during manipulation. In all of these cases the secret has a high likelihood of ending up in the memory dump.
+
+That secrets originate as strings deserves a bit more attention. The secret almost certainly entered the application via an HTTP request, or it was read from config, or it was entered into a WinForms textbox. These copies of the secrets are still floating around in memory (potentially even in a shared pool!), and SecureString does nothing to mitigate their being leaked.
+
+The only reliable mechanism for preventing secrets from getting into crash dumps is to disable crash dumps for applications which process sensitive data. For web servers, this might mean separating out the "sensitive" parts of the application (login, payment processing, etc.) into a separate service running in a separate secured environment. The remaining aspects of the web service can then run with normal crash dump diagnostics provided they never have access to the plaintext form of the sensitive data.
```

Comment:
Have you considered `MADV_DONTDUMP` ? Others are using it for this purpose, e.g. here: https://jira.mongodb.org/browse/SERVER-21503

As long as the dump collector is compliant and the process is not hacked, it should prevent secrets in `MADV_DONTDUMP` pages from leaking into crash dump.

Is there an equivalent on Windows?

Reply:
Good idea - I'll check. The current prototype (on Windows) uses `HeapCreate` to carve out a separate region of the heap for storage of this data. This wasn't intended to hide the data from dump collection. Rather, the intent was to reduce the odds that some dangling `malloc` pointer might point to this region of memory.

Reply:
>  Presumably the app would need to use the secret at some point; then the secret would need to be moved to the normal heap

If there is a decent interop support for this, the app does not need to move the secret out into regular heap. The secret can always live in the special heap only and anything that needs interop with it would use the pointer into the special memory region or create a copy in the special memory region.

Reply:
In theory we could provide a span-based callback, which would allow something akin to _String.Create_.

```cs
var shroudedData = /* ... */;
shroudedData.Use(state, (var span, var state) => { /* use 'span' here */ });
```

This would allow the shrouded data instance to find an appropriate place to put the secret and to pass that to the lambda function. But it has the same drawbacks as _String.Create_: your state object cannot itself contain spans, so you can't nest these things.

At a lower level we could expose a handle to the secret, but this now means that the secret has to exist in some form in readable memory. The early experiments I did with this type either encrypted the memory in-proc or shuttled the secret out-of-process entirely (to lsass), neither of which would allow for direct pointer manipulation.

Reply:
> Have you considered `MADV_DONTDUMP`?

Yes. The Windows equivalent is [`WerRegisterExcludedMemoryBlock`](https://docs.microsoft.com/windows/win32/api/werapi/nf-werapi-werregisterexcludedmemoryblock). However, per the documentation for both functions, they're not intended to prevent the capture of secret data. They're instead intended to prevent the capture of extremely large regions of no value from crash dumps. The canonical example is texture map data for games, as these maps can potentially be gigabytes in size and will almost never be relevant for debugging crash dumps. On Windows there's an upper bound to how many entries each process can have in the exclusion list. It's not advised to use these functions to mark lots of small regions as "no dump", because it could end up exhausting entries and prevent the application from registering a legitimate single large region.